sycl/doc/extensions/proposed/sycl_ext_oneapi_user_defined_reductions.asciidoc

```diff
@@ -0,0 +1,220 @@
+
+https://github.com/intel/llvm/issues
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 5 specification. All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../experimental/sycl_ext_oneapi_group_sort.asciidoc[
+  sycl_ext_oneapi_group_sort]
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+== Overview
+
+The purpose of this extension is to expand functionality of `reduce_over_group`
+and `joint_reduce` free functions defined in section 4.17.4.5. `reduce` of the
+core SYCL specification by allowing user-defined binary operators and
+non-fundamental types.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 Feature test macros. Therefore, an implementation
+supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS` to one of the values defined in the
+table below.
+Application can test for existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's value
+to determine which of the extensions's APIs the implementation supports.
+
+Table 1. Values of the `SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS` macro.
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+=== Reduction functions
+
+This extension provides two overloads of `reduce_over_group` defined by the core
+SYCL specification.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <typename GroupHelper, typename Ptr, typename BinaryOperation>
+  std::iterator_traits<Ptr>::value_type joint_reduce(GroupHelper g, Ptr first, Ptr last, BinaryOperation binary_op); // (1)
+
+  template <typename GroupHelper, typename Ptr, typename T, typename BinaryOperation>
+  T joint_reduce(GroupHelper g, Ptr first, Ptr last, T init, BinaryOperation binary_op); // (2)
+
+  template <typename GroupHelper, typename T, typename BinaryOperation>
+  T reduce_over_group(GroupHelper g, T x, BinaryOperation binary_op); // (3)
+
+  template <typename GroupHelper, typename V, typename T, typename BinaryOperation>
+  T reduce_over_group(GroupHelper g, V x, T init, BinaryOperation binary_op); // (4)
+}
+----
+
+**TODO**: consider adding `is_group_with_scratchpad_v` helper to describe
+preconditions better.
+
+1._Constraints_: `GroupHelper` is a `group_with_scratchpad`.
```

Comment:
Why is it required for this parameter to be specifically `group_with_scratchpad` rather than any object that satisfies [Group Helper](https://github.com/intel/llvm/blob/e33b270f6a89976ed1642f5b3daee7721197af83/sycl/doc/extensions/experimental/sycl_ext_oneapi_group_sort.asciidoc#group-helper)?

I actually forget why the sycl_ext_oneapi_group_sort extension defines a general concept of Group Helper rather than only the concrete type `group_with_scratchpad`.  However, it seems like these two extensions should be aligned about whether this parameter must be the concrete type `group_with_scratchpad` vs. any object that satisfies Group Helper.

Reply:
I think it is just me missing the fact that we are introducing a new named interface/contract. At least in this extension there should be no problems with accepting a general Group Helper instead of particular `group_with_scratchpad` object.

My preference would be if we had a `is_group_helper`/`is_group_with_scratchpad` type trait, which we could use in both specification and implementation.

Reply:
Wouldn't it be inconvenient to the user if we added a trait like `is_group_helper`?  If a user defines their own type that satisfies the requirements of "Group Helper", would they also be required to specialize the trait for that type?

Reply:
Can you work with @andreyfe1 to decide what makes sense, so we have consistency between this extension and "sycl_ext_oneapi_group_sort"?  From my point of view, any of the following would be OK:

* We remove the flexibility and decide that the concrete type `group_with_scratchpad` is enough for now.
* We allow any type that satisfies "Group Helper", add a trait like `is_group_helper`, but don't allow users to define their own types.
* We allow any type that satisfies "Group Helper", add a trait like `is_group_helper`, and somehow set that trait to `true` automatically for user-defined types.
* We allow any type that satisfies "Group Helper", do not add a trait, but add some static_assert that gives a clear error message if the user passes a type that does not satisfy "Group Helper".

Tagging @Pennycook also since he may have opinions here.

Reply:
Synced offline with @andreyfe1, Andrei suggested 

1. to go with option 2 in the near future:

   * We allow any type that satisfies "Group Helper", add a trait like is_group_helper, but don't allow users to define their own types.

2. Move `GroupHelper` to separate extension as this is currently used by multiple extensions.

Reply:
Since this extension is documented to depend on [sycl_ext_oneapi_group_sort](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc), I think we could add `is_group_helper` to sycl_ext_oneapi_group_sort rather than creating a separate extension.

I prefer to avoid adding extensions that are not usable on their own.  An extension that only defines `is_group_helper` would not be usable on its own.

Reply:
Thanks, applied in https://github.com/intel/llvm/pull/7202/commits/26acc0117b4f1fc3ffe17baf330d90ddbc3936de and https://github.com/intel/llvm/pull/7202/commits/6475c7504767ceb16f8347d2e5eb7c7a3decc0ef